### PR TITLE
modified all subclasses of FilesBaseTest to use real filesystem file

### LINKED ISSUE #2337: Avoid mocks for testing `File` assertions
Similar to #2292, all the test cases in subclasses of `FilesBaseTest` should use real filesystem elements and not mocks.

We might still favor the NIO2 File API to initialize the test prerequisites, so `actual` would be obtained via `Path::toFile`.

### PR DESCRIPTION
**Pull request by ahmadsadeed**
#### Check List:
* Fixes #2337 
* Unit tests : NA
* Javadoc with a code example (on API only) : NA
* PR meets the [contributing guidelines](https://github.com/assertj/assertj-core/blob/main/CONTRIBUTING.md)

This is my first PR on Assertj. I will appreciate if I can get any suggestions on how to improve my work for this and future issues.

### CODE REVIEW

File: src/test/java/org/assertj/core/internal/files/File_assertHasSize_Test.java [hunk 1]
```diff
@@ -68,8 +69,9 @@ void should_fail_if_actual_is_not_a_file() {
     AssertionInfo info = someInfo();
     File notAFile = new File("xyz");
     // WHEN
-    expectAssertionError(() -> files.assertHasSizeInBytes(info, notAFile, 36L));
+    AssertionError error = expectAssertionError(() -> files.assertHasSizeInBytes(info, notAFile, 36L));
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/internal/files/File_assertHasSize_Test.java:72]: `expectAssertionError` already makes sure the error is an `AssertionError`, please revert all the changes in this file.
The only thing to do is remove @DisplayName("Files.assertHasSizeInBytes:") as it is computed automatically

Comment (joel-costigliola) [reply src/test/java/org/assertj/core/internal/files/File_assertHasSize_Test.java:72]: this comment applies for the other tests too ;-)

File: src/test/java/org/assertj/core/internal/files/Files_assertHasBinaryContent_Test.java [hunk 2]
```diff
@@ -94,13 +108,20 @@ void should_throw_error_wrapping_caught_IOException() throws IOException {
...
+    byte[] expected = "fake".getBytes();
+
+    BinaryDiff binaryDiff = new BinaryDiff();
+    BinaryDiffResult diff = binaryDiff.diff(actual, expected);
     AssertionInfo info = someInfo();
 
-    Throwable error = catchThrowable(() -> files.assertHasBinaryContent(info, actual, expected));
+    AssertionError error = expectAssertionError(() -> unMockedFiles.assertHasBinaryContent(info, actual, expected));
 
     assertThat(error).isInstanceOf(AssertionError.class);
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/internal/files/Files_assertHasBinaryContent_Test.java:124]: can be removed as you are using `expectAssertionError`

File: src/test/java/org/assertj/core/internal/files/Files_assertHasBinaryContent_Test.java [hunk 3]
```diff
@@ -94,13 +108,20 @@ void should_throw_error_wrapping_caught_IOException() throws IOException {
...
-    verify(failures).failure(info, shouldHaveBinaryContent(actual, diff));
+    then(error).hasMessage(shouldHaveBinaryContent(actual, diff).create(info.description(), info.representation()));
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/internal/files/Files_assertHasBinaryContent_Test.java:125]: Revert this change as the previous code basically does the same thing but in a more concise way

Comment (joel-costigliola) [reply src/test/java/org/assertj/core/internal/files/Files_assertHasBinaryContent_Test.java:125]: this comment applies to other tests too

File: src/test/java/org/assertj/core/internal/files/Files_assertHasBinaryContent_Test.java [hunk 4]
```diff
@@ -94,13 +108,20 @@ void should_throw_error_wrapping_caught_IOException() throws IOException {
 
   @Test
   void should_fail_if_file_does_not_have_expected_binary_content() throws IOException {
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/internal/files/Files_assertHasBinaryContent_Test.java:110]: add GIVEN WHEN THEN to the tests you are updating, thanks!

Comment (joel-costigliola) [reply src/test/java/org/assertj/core/internal/files/Files_assertHasBinaryContent_Test.java:110]: this is a general comment that applies to other tests

**Review by joel-costigliola — CHANGES_REQUESTED** [review]
Thanks , first round of code review done. Thanks @ahmadsadeed!

**Comment by ahmadsadeed** [comment]
Thanks @joel-costigliola, I made the requested changes. I added `protected Failures unMockedFailures;` to FilesBaseTest.java since I was using `unMockedFiles` instead of `files` in some tests. Then used `verify(unMockedFailures)` instead of `verify(failures)`. I hope I did this part right.

**Comment by ahmadsadeed** [comment]
By the way, I noticed that the CI workflow was failing due to an issue in Files_assertCanRead_Test. I fixed this issue so it shouldn't be a problem.

File: src/test/java/org/assertj/core/internal/files/Files_assertCanRead_Test.java [hunk 5]
```diff
@@ -46,18 +42,19 @@ void should_fail_if_actual_is_null() {
 
   @Test
   void should_fail_if_can_not_read() {
-    when(actual.canRead()).thenReturn(false);
+    // GIVEN
+    File notAFile = new File("xyz");
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/internal/files/Files_assertCanRead_Test.java:46]: minor: rename to `nonExistentFile` which is more accurate as `notAFile` is a `File` but simply not an existing one

File: src/test/java/org/assertj/core/internal/files/Files_assertCanWrite_Test.java [hunk 6]
```diff
@@ -46,18 +43,19 @@ void should_fail_if_actual_is_null() {
...
-
+    // THEN
     assertThat(error).isInstanceOf(AssertionError.class);
     verify(failures).failure(info, shouldBeWritable(actual));
   }
 
   @Test
   void should_pass_if_actual_can_write() {
-    when(actual.canWrite()).thenReturn(true);
+    File actual = newFile(tempDir.getAbsolutePath() + "to_write.txt");
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/internal/files/Files_assertCanWrite_Test.java:58]: Even if it's experimental let's give a try to JUnit5 [@TempDir](https://junit.org/junit5/docs/current/user-guide/#writing-tests-built-in-extensions-TempDirectory) to avoid leaving temp files.

Have a look at `PathsBaseTest` where it is already used.

Comment (joel-costigliola) [reply src/test/java/org/assertj/core/internal/files/Files_assertCanWrite_Test.java:58]: Comment applies to other tests using temp files

File: src/test/java/org/assertj/core/internal/files/Files_assertHasBinaryContent_Test.java [hunk 7]
```diff
@@ -66,19 +67,25 @@ void should_fail_if_actual_is_null() {
...
   void should_pass_if_file_has_expected_binary_content() throws IOException {
-    when(binaryDiff.diff(actual, expected)).thenReturn(BinaryDiffResult.noDiff());
-    files.assertHasBinaryContent(someInfo(), actual, expected);
+    File actual = newFile(tempDir.getAbsolutePath() + "/tmp.txt");
+    byte[] data = "actual".getBytes();
+    try (FileOutputStream myWriter = new FileOutputStream(actual)) {
+      myWriter.write(data, 0, data.length);
+    }
+    byte[] expected = "actual".getBytes();
+    unMockedFiles.assertHasBinaryContent(someInfo(), actual, expected);
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/internal/files/Files_assertHasBinaryContent_Test.java:88]: Take an approach similar to `Paths_assertHasBinaryContent_Test`, it's using a more concise way of writing files:
```
// call .toFile() other it returns a Path
Files.write(tempDir.resolve("actual"), "Content".getBytes()).toFile();
```

Comment (joel-costigliola) [reply src/test/java/org/assertj/core/internal/files/Files_assertHasBinaryContent_Test.java:88]: This comment applies to other tests too (I mean looking at the equivalent test for Path)

**Review by joel-costigliola — COMMENTED** [review]
thanks for the PR @ahmadsadeed, this goes in the right direction.
I haven't commented everything to avoid repeating myself too much, the comments I made apply to many tests, if can you address them  to all the tests where they make sense that would be fantastic.

File: src/test/java/org/assertj/core/internal/files/Files_assertHasBinaryContent_Test.java (hunk 7 above)
Comment (ahmadsadeed) [reply src/test/java/org/assertj/core/internal/files/Files_assertHasBinaryContent_Test.java]: Since the two unresolved comments are on the same topic, I will ask my question here:

- We already have `import org.assertj.core.internal.Files;` so I have to use `java.nio.file.Files.write`. Any suggestions or comments?
- `tmpDir` is declared as type `protected File tempDir;` in FilesBaseTest. I could use `tempDir.toPath()` or change it to `protected Path tempDir;`. Or keep both types:
```
  @TempDir
  protected File tempDir;
  @TempDir
  protected Path tempDirPath;
```
Final:
```
  @Test
  void should_throw_error_if_actual_file_does_not_have_the_expected_size() throws IOException {
    // GIVEN
    File actual = java.nio.file.Files.write(tempDirPath.resolve("actual"), "content".getBytes()).toFile();
    // WHEN
    expectAssertionError(() -> files.assertHasSizeInBytes(INFO, actual, 36L));
    // THEN
    verify(failures).failure(INFO, shouldHaveSize(actual, 36L));
  }
```

Comment (ahmadsadeed) [reply src/test/java/org/assertj/core/internal/files/Files_assertHasBinaryContent_Test.java]: Also, should I replace usage of files from resources folder with the suggestion above?
```@BeforeAll
  static void setUpOnce() {
    actual = new File("src/test/resources/actual_file.txt");
  }
```

```@Test
  void should_pass_if_actual_has_expected_size() {
    files.assertHasSizeInBytes(INFO, actual, actual.length());
  }
```

Comment (joel-costigliola) [reply src/test/java/org/assertj/core/internal/files/Files_assertHasBinaryContent_Test.java]: For files test let's use:
```
  @TempDir
  protected File tempDir;
```
unless I'm missing something we don't need a `Path tempDir` .

Comment (joel-costigliola) [reply src/test/java/org/assertj/core/internal/files/Files_assertHasBinaryContent_Test.java]: we could add a helper method in `FilesBaseTest`  to wrap:
```java
java.nio.file.Files.write(tempDirPath.resolve("actual"), "content".getBytes()).toFile();
```
to something like:
```
writeFile("actual", "content");
```
That's a suggestion, maybe try it and see for yourself it it makes it easier otherwise don't worry too much ;-)

File: src/test/java/org/assertj/core/internal/files/Files_assertHasBinaryContent_Test.java [hunk 8]
```diff
@@ -54,53 +55,64 @@ static void setUpOnce() {
...
 
   @Test
   void should_pass_if_file_has_expected_binary_content() throws IOException {
-    when(binaryDiff.diff(actual, expected)).thenReturn(BinaryDiffResult.noDiff());
-    files.assertHasBinaryContent(someInfo(), actual, expected);
+    File actual = newFile(tempDir.getAbsolutePath() + "/tmp.txt");
+    byte[] data = "actual".getBytes();
+    try (FileOutputStream myWriter = new FileOutputStream(actual)) {
+      myWriter.write(data, 0, data.length);
+    }
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/internal/files/Files_assertHasBinaryContent_Test.java:84]: this comment applies broadly: use [commons io FileUtils](https://commons.apache.org/proper/commons-io/apidocs/org/apache/commons/io/FileUtils.html) to write file content, there are many ways, ex `writeByteArrayToFileFile file, byte[] data)` or `write(File file, CharSequence data, String charsetName)`. These methods throw `IOException`, don't catch them just add them in the test method throw declaration.

**Review by joel-costigliola — CHANGES_REQUESTED** [review]
@ahmadsadeed thanks this looks good!
I added a comment to use commons io to simplify writing files content.

The other one is optional (I'll do it if you don't have time) is to refactor the `should_fail_if_actual_is_null()` tests to be like:

```java
  @Test
  void should_fail_if_actual_is_null() {
    // GIVEN
    File actual = null;
    // WHEN
    AssertionError error = expectAssertionError(() -> files.assertSameBinaryContentAs(INFO, actual, expected));
    // THEN
    then(error).hasMessage(actualIsNull());
  } 
```
After that we are good to go!

**Comment by joel-costigliola** [comment]
Integrated thanks @ahmadsadeed!
